operator/pkg/ingress/ingress_class.go

```diff
@@ -4,172 +4,199 @@
+		}
 	}
-	return err
 }
 
-func (i *ingressClassManager) handleIngressClassAddedEvent(event ingressClassAddedEvent) error {
-	log.WithField(logfields.IngressClass, event.ingressClass).Debug("Handling ingress class add")
-	i.notify(event.ingressClass)
-	return nil
-}
+// handleSyncEvent handles a Sync event for the events of an IngressClasses Resource.
+func (i *ingressClassManager) handleSyncEvent() error {
+	log.Debug("Handling IngressClass Sync event")
+	i.synced.Store(true)
 
-func (i *ingressClassManager) handleIngressClassUpdatedEvent(event ingressClassUpdatedEvent) error {
-	log.WithField("old", event.oldIngressClass).WithField("new", event.newIngressClass).Debug("Handling ingress class update")
-	i.notify(event.newIngressClass)
 	return nil
 }
 
-func (i *ingressClassManager) handleIngressClassDeletedEvent(event ingressClassDeletedEvent) error {
-	log.WithField(logfields.IngressClass, event.ingressClass).Debug("Handling ingress class delete")
+// handleUpsertEvent handles an Upsert event for an IngressClass resource.
+// If the IngressClass has a name that is not the Cilium IngressClass name, it will be ignored.
+// If an error is returned, then the IngressClass contained in the event should be retried at a later
+// point in time.
+func (i *ingressClassManager) handleUpsertEvent(event resource.Event[*slim_networkingv1.IngressClass]) error {
+	log.WithField(logfields.IngressClass, event.Object).Warn("Handling IngressClass Upsert event")
 
-	if event.ingressClass.GetName() == ciliumIngressClassName {
-		i.ingressQueue.Add(ciliumIngressClassDeletedEvent(event))
+	if event.Object == nil || (event.Object != nil && !isIngressClassCilium(event.Object)) {
+		return nil
 	}
-	return nil
+
+	// Even if an error occurs due to a bad annotation value, we still need to perform an update to
+	// signal that we are no longer the default ingress class.
+	isDefault, err := isIngressClassMarkedDefault(event.Object)
+	old := i.isDefaultIngressClass.Swap(isDefault)
+
+	i.queue.Add(ciliumIngressClassUpdatedEvent{
+		isDefault: isDefault,
+		changed:   !(old == isDefault),
+	})
+
+	return err
 }
 
-// notify informs parent ingress about change in our ingress class configuration
-func (i *ingressClassManager) notify(ic *slim_networkingv1.IngressClass) {
-	if ic.GetName() == ciliumIngressClassName {
-		i.ingressQueue.Add(ciliumIngressClassUpdatedEvent{ingressClass: ic})
+// handleDeleteEvent handles a Delete event for an IngressClass resource.
+// If the IngressClass has a name that is not the Cilium IngressClass name, it will be ignored.
+// If an error is returned, then the IngressClass contained in the event should be retried at a later
+// point in time.
+func (i *ingressClassManager) handleDeleteEvent(event resource.Event[*slim_networkingv1.IngressClass]) error {
+	log.WithField(logfields.IngressClass, event.Object).Debug("Handling IngressClass Delete event")
+
+	if event.Object == nil || (event.Object != nil && !isIngressClassCilium(event.Object)) {
```

Comment:
same here: no need to check `event.Object` being not nil in the second part of the condition.

-> `if event.Object == nil || !isIngressClassCilium(event.Object) {`